### ISSUE TITLE
panic: send on closed channel

### ISSUE THREAD

**Issue by chrsmith** [pulumi/pulumi#2323]
We've started to see failures like this in our CI/CD.

I would guess this is related to #2314 , where we start recording engine events concurrently. If the channel processing all of the diagnostic events is closed before those concurrent events are processed, they would send on a closed channel and panic. Not sure how the `jaeger-client-go` package is involved, but the interesting part of the callstack starts at `httpstate/client.(*Client).RecordEngineEvent`.

```
[ /examples/aws-ts-ruby-on-rails ] panic: send on closed channel
[ /examples/aws-ts-ruby-on-rails ] 
[ /examples/aws-ts-ruby-on-rails ] goroutine 425 [running]:
[ /examples/aws-ts-ruby-on-rails ] github.com/pulumi/pulumi/vendor/github.com/uber/jaeger-client-go.(*remoteReporter).Report(0xc0000b93e0, 0xc000412ff0)
[ /examples/aws-ts-ruby-on-rails ] 	/home/travis/gopath/src/github.com/pulumi/pulumi/vendor/github.com/uber/jaeger-client-go/reporter.go:209 +0x47
[ /examples/aws-ts-ruby-on-rails ] github.com/pulumi/pulumi/vendor/github.com/uber/jaeger-client-go.(*Tracer).reportSpan(0xc00030a3c0, 0xc000412ff0)
[ /examples/aws-ts-ruby-on-rails ] 	/home/travis/gopath/src/github.com/pulumi/pulumi/vendor/github.com/uber/jaeger-client-go/tracer.go:356 +0xd7
[ /examples/aws-ts-ruby-on-rails ] github.com/pulumi/pulumi/vendor/github.com/uber/jaeger-client-go.(*Span).FinishWithOptions(0xc000412ff0, 0xbf004236566d1c0f, 0x29e1fe1e5, 0x1969820, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
[ /examples/aws-ts-ruby-on-rails ] 	/home/travis/gopath/src/github.com/pulumi/pulumi/vendor/github.com/uber/jaeger-client-go/span.go:217 +0x439
[ /examples/aws-ts-ruby-on-rails ] github.com/pulumi/pulumi/vendor/github.com/uber/jaeger-client-go.(*Span).Finish(0xc000412ff0)
[ /examples/aws-ts-ruby-on-rails ] 	/home/travis/gopath/src/github.com/pulumi/pulumi/vendor/github.com/uber/jaeger-client-go/span.go:195 +0x67
[ /examples/aws-ts-ruby-on-rails ] github.com/pulumi/pulumi/pkg/backend/httpstate/client.pulumiAPICall(0xff4a00, 0xc0006c6e70, 0xff91a0, 0xc0002999b0, 0xc0000b59c0, 0x16, 0xeb2a35, 0x4, 0xc0001c5880, 0x6a, ...)
[ /examples/aws-ts-ruby-on-rails ] 	/home/travis/gopath/src/github.com/pulumi/pulumi/pkg/backend/httpstate/client/api.go:234 +0x129f
[ /examples/aws-ts-ruby-on-rails ] github.com/pulumi/pulumi/pkg/backend/httpstate/client.pulumiRESTCall(0xff4a00, 0xc0006c6e70, 0xff91a0, 0xc0002999b0, 0xc0000b59c0, 0x16, 0xeb2a35, 0x4, 0xc0001c5880, 0x6a, ...)
[ /examples/aws-ts-ruby-on-rails ] 	/home/travis/gopath/src/github.com/pulumi/pulumi/pkg/backend/httpstate/client/api.go:268 +0x180
[ /examples/aws-ts-ruby-on-rails ] github.com/pulumi/pulumi/pkg/backend/httpstate/client.(*Client).updateRESTCall(0xc00008e3c0, 0xff4a00, 0xc0006c6e70, 0xeb2a35, 0x4, 0xc0001c5880, 0x6a, 0x0, 0x0, 0xe75760, ...)
[ /examples/aws-ts-ruby-on-rails ] 	/home/travis/gopath/src/github.com/pulumi/pulumi/pkg/backend/httpstate/client/client.go:82 +0x1af
[ /examples/aws-ts-ruby-on-rails ] github.com/pulumi/pulumi/pkg/backend/httpstate/client.(*Client).RecordEngineEvent(0xc00008e3c0, 0xff4a00, 0xc0006c6e70, 0xc000eba180, 0x7, 0xc00003e8a0, 0x23, 0xeb6a87, 0x7, 0xc0001096e0, ...)
[ /examples/aws-ts-ruby-on-rails ] 	/home/travis/gopath/src/github.com/pulumi/pulumi/pkg/backend/httpstate/client/client.go:514 +0x19d
[ /examples/aws-ts-ruby-on-rails ] github.com/pulumi/pulumi/pkg/backend/httpstate.(*cloudUpdate).recordEngineEvent(0xc00014cb80, 0xeb4d06, 0x6, 0x0, 0x0, 0x24, 0x0, 0x0)
[ /examples/aws-ts-ruby-on-rails ] 	/home/travis/gopath/src/github.com/pulumi/pulumi/pkg/backend/httpstate/state.go:159 +0x2b1
[ /examples/aws-ts-ruby-on-rails ] github.com/pulumi/pulumi/pkg/backend/httpstate.(*cloudUpdate).RecordAndDisplayEvents.func2(0xeb4d06, 0x6, 0x0, 0x0, 0x24)
[ /examples/aws-ts-ruby-on-rails ] 	/home/travis/gopath/src/github.com/pulumi/pulumi/pkg/backend/httpstate/state.go:196 +0x8f
[ /examples/aws-ts-ruby-on-rails ] created by github.com/pulumi/pulumi/pkg/backend/httpstate.(*cloudUpdate).RecordAndDisplayEvents
[ /examples/aws-ts-ruby-on-rails ] 	/home/travis/gopath/src/github.com/pulumi/pulumi/pkg/backend/httpstate/state.go:207 +0x35a
[ /examples/aws-ts-ruby-on-rails ] 
```

**Comment by joeduffy** [comment]
Are you running with the fix to https://github.com/pulumi/pulumi/issues/2170?

**Comment by chrsmith** [comment]
Yeah, looks like this was fixed in v2.11.0 of the library, and we are still on v2.9.0.
https://github.com/jaegertracing/jaeger-client-go/releases/tag/v2.11.1

- https://github.com/jaegertracing/jaeger-client-go/pull/235
- https://github.com/jaegertracing/jaeger-client-go/pull/245

**Comment by lukehoban** [comment]
Note that we thought we solved the underlying Jaeger client error in https://github.com/pulumi/pulumi/pull/2290 which “fixed” #1850. 

Seems like some more recent change re-agravated that, as current master is now hitting this on almost every test.

**Comment by chrsmith** [comment]
Closing this out since things are green again. The Jaeger client issue only happens when there is a bug on the consuming end. (Which evidently we had 😢 .) #2329 will entirely eliminate the possibility, but for the time being there is no need to have this bug open.
